src/ModelContextProtocol/Server/McpServer.cs

```diff
@@ -401,7 +414,19 @@ private void SetToolsHandler(McpServerOptions options)
                     do
                     {
                         ListToolsResult extraResults = await originalListToolsHandler(request, cancellationToken).ConfigureAwait(false);
-                        result.Tools.AddRange(extraResults.Tools);
+                        foreach(var tool in extraResults.Tools)
+                        {
+                            resultTools.Add(tool);
+                            if (progressToken is not null)
+                            {
+                                await this.NotifyProgressAsync(new()
+                                {
+                                    ProgressToken = progressToken,
+                                    Progress = resultTools.Count,
+                                    Message = $"Listing extra tool '{tool.Name}'...",
```

Comment:
The client shouldn't need any information about how the server is implemented, whereas having this message differ from the one above leaks details about how the server is structured. We should ensure the messages are effectively the same.